R-package/tests/testthat/test_lgb.Booster.R

```diff
@@ -1518,3 +1518,73 @@ test_that("LGBM_BoosterGetNumFeature_R returns correct outputs", {
     ncols <- .Call(LGBM_BoosterGetNumFeature_R, model$.__enclos_env__$private$handle)
     expect_equal(ncols, ncol(iris) - 1L)
 })
+
+set.seed(1L)
+data(agaricus.train, package = "lightgbm")
+train <- agaricus.train
+bst <- lightgbm(
+  data = as.matrix(train$data)
+  , label = train$label
+  , params = list(objective = "binary", num_threads = .LGB_MAX_THREADS)
+  , nrounds = 5L
+  , verbose = .LGB_VERBOSITY
+)
+
+# Simplified version of lgb.model.dt.tree()
+get_trees_from_dump <- function(x) {
+  parsed <- jsonlite::fromJSON(
+    txt = x
+    , simplifyVector = TRUE
+    , simplifyDataFrame = FALSE
+    , simplifyMatrix = FALSE
+    , flatten = FALSE
+  )
+  return(lapply(parsed$tree_info, FUN = .single_tree_parse))
+}
+
+get_n_trees <- function(x) {
+  return(length(get_trees_from_dump(lgb.dump(x))))
+}
+
+test_that("num_iteration and start_iteration work for lgb.dump()", {
+  first2 <- get_trees_from_dump(lgb.dump(bst, num_iteration = 2L))
+  last3 <- get_trees_from_dump(
+    lgb.dump(bst, num_iteration = 3L, start_iteration = 2L)
+  )
+  all5 <- get_trees_from_dump(lgb.dump(bst))
+  too_many <- get_trees_from_dump(lgb.dump(bst, num_iteration = 10L))
+
+  expect_equal(c(first2, last3), all5)
+  expect_equal(too_many, all5)
```

Comment:
The `dump_model()` and `lgb.model.dt.tree()` output contain a `tree_index`, a unique identifier for the **exact** tree that's represented.

To catch issues like ordering, NULLs, etc., can you also please test for the inclusion of the **specific, expected `tree_index` values**?

For example, with `num_iteration=3L, start_iteration = 2L`, I expect to see `tree_index` values of `{2, 3, 4}` represented in the output.

Here's an example:

```r
library(lightgbm)
data(agaricus.train, package = "lightgbm")
train <- agaricus.train
bst <- lightgbm(
  data = as.matrix(train$data)
  , label = train$label
  , params = list(objective = "binary")
  , nrounds = 5L
)

parsed <- jsonlite::fromJSON(
    txt = bst$dump_model()
    , simplifyVector = TRUE
    , simplifyDataFrame = FALSE
    , simplifyMatrix = FALSE
    , flatten = FALSE
)

parsed[["tree_info"]][[1L]][["tree_index"]]
# 0

parsed[["tree_info"]][[2L]][["tree_index"]]
# 1
```

Reply:
I have now changed this test to compare the full results.